### PR TITLE
Add pipeline metrics

### PR DESCRIPTION
**Pull request by FedericoCeratto**
Add metric support to centrifugation
Improve test_integration.py: support configurable docker image tag, delete old Postgres container
Test if docker pings
Add docstrings
General refactor
Add tox.ini

### CODE REVIEW

File: af/shovel/Dockerfile [hunk 1]
```diff
@@ -5,7 +5,7 @@ FROM ubuntu:16.04
 # All the actions are done within single RUN to reduce number of layers & overall fetch size.
 RUN set -ex \
     && apt-get update \
-    && apt-get install -y python2.7 liblz4-tool python-ujson python-yaml python-psycopg2 python-six \
+    && apt-get install -y python2.7 liblz4-tool python-ujson python-yaml python-psycopg2 python-six python-statsd\
```
Comment (bassosimone) [inline af/shovel/Dockerfile]: Nit: maybe keep a whitespace before the `\`

File: af/shovel/centrifugation.py [hunk 2]
```diff
@@ -34,6 +34,10 @@
 from oonipl.sim import sim_shi4_mm3_layout, sim_shi4_mm3_text
 from oonipl.utils import dclass
 
+from oonipl.metrics import setup_metrics
+
+metrics = setup_metrics(host="172.17.0.1", name="centrifugation")
```
Comment (bassosimone) [inline af/shovel/centrifugation.py:40]: I believe we should probably document what this IP address is. I cannot figure out immediately what is its purpose, but I believe the `172.17` space is used by Docker, right?

File: af/shovel/oonipl/metrics.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+"""
+
+from os.path import basename, splitext
+
+import statsd
+
+
+def setup_metrics(host="localhost", name=None):
+    """Setup metric generation. Use dotted namespaces e.g.
+    "pipeline.centrifugation"
```
Comment (bassosimone) [inline af/shovel/oonipl/metrics.py:14]: Maybe you can `assert` or `throw` if a dotted namespace is important?

File: tests/requirements.txt [hunk 4]
```diff
@@ -1,3 +1,4 @@
+# Required by tests/test_integration.py
```
Comment (bassosimone) [inline tests/requirements.txt:1]: 👍

File: tests/test_integration.py [hunk 5]
```diff
@@ -14,39 +21,54 @@
...
+        running[0].stop()
+        running[0].remove()
+
+    print("Starting Postgres container")
+    pg_container = docker_client.containers.run(
+        "postgres:9.6",
+        name=METADB_NAME,
+        hostname=METADB_NAME,
+        environment={"POSTGRES_USER": METADB_PG_USER},
+        ports={"5432/tcp": 25432},
```
Comment (bassosimone) [inline tests/test_integration.py:44]: This is the equivalent of `docker run --publish ...` right?

File: tests/test_integration.py [hunk 6]
```diff
@@ -14,39 +21,54 @@
...
+        detach=True,
     )
     while True:
```
Comment (bassosimone) [inline tests/test_integration.py:48]: Meta thought: do we care about this potentially running forever?

File: tests/test_integration.py [hunk 7]
```diff
@@ -14,39 +21,54 @@
...
 
+
 def pg_install_tables(container):
-    _, socket = container.exec_run(cmd="psql -U {}".format(METADB_PG_USER), stdin=True, socket=True)
-    if hasattr(socket, '_sock'):
+    _, socket = container.exec_run(
+        cmd="psql -U {}".format(METADB_PG_USER), stdin=True, socket=True
+    )
+    if hasattr(socket, "_sock"):
         socket = socket._sock
```
Comment (bassosimone) [inline tests/test_integration.py:71]: Why is this socket unwrapping necessary? (I think you probably didn't write this code but I am curious to know why accessing `_sock` is necessary in this context and maybe you know that.)

File: tests/test_integration.py [hunk 8]
```diff
@@ -55,90 +77,100 @@ def pg_install_tables(container):
...
-                'bind': '/mnt',
-                'mode': 'rw'
-            }
-        },
+        mem_limit="512m",
+        user="{}:{}".format(os.getuid(), os.getgid()),
+        working_dir="/mnt",
+        links={METADB_NAME: None},
+        volumes={REPO_ROOT: {"bind": "/mnt", "mode": "rw"}},
         privileged=True,
```
Comment (bassosimone) [inline tests/test_integration.py]: Do you know why this container needs to be `--privileged`?

File: tests/test_integration.py [hunk 9]
```diff
@@ -55,90 +77,100 @@ def pg_install_tables(container):
...
+        sys.stdout.write(line.decode("utf8"))
         sys.stdout.flush()
     print("runtime: {}".format(time.time() - start_time))
     return shovel_container
 
+
 def pg_conn():
-    return psycopg2.connect('host={} user={} port={}'.format('localhost', METADB_PG_USER, 25432))
+    return psycopg2.connect(
+        "host={} user={} port={}".format("localhost", METADB_PG_USER, 25432)
```
Comment (bassosimone) [inline tests/test_integration.py]: I'd consider whether we can avoid hardcoding the port number

**Review by bassosimone — CHANGES_REQUESTED** [review]
I was able to understand the diff and it mostly looks good to me. I have asked a number of clarifying questions and provided some suggestions. The most pressing concern at the moment is that the CI is failing with multiple errors. My understanding is that somehow `statsd` is missing.

File: af/shovel/oonipl/metrics.py (hunk 3 above)
Comment (FedericoCeratto) [reply af/shovel/oonipl/metrics.py:14]: The dotted structure is for our convenience to create a hierarchy of metrics. If a component is clearly identified by a single word we can use it without dots.

**Comment by FedericoCeratto** [comment]
Example of metrics (displayed by Netdata). The ingestion speed is the most interesting:

image
image

File: af/shovel/Dockerfile (hunk 1 above)
Comment (FedericoCeratto) [reply af/shovel/Dockerfile]: Done

File: af/shovel/centrifugation.py (hunk 2 above)
Comment (FedericoCeratto) [reply af/shovel/centrifugation.py:40]: Updated

File: tests/test_integration.py (hunk 6 above)
Comment (FedericoCeratto) [reply tests/test_integration.py:48]: It hasn't been a problem so far but a better try/finally to stop it even on test failure would be good.

File: tests/test_integration.py (hunk 9 above)
Comment (FedericoCeratto) [reply tests/test_integration.py]: It appears only in 2 places and this is just a test harness...

File: tests/test_integration.py (hunk 8 above)
Comment (FedericoCeratto) [reply tests/test_integration.py]: Unsure. It seems to work without it.

File: tests/test_integration.py (hunk 7 above)
Comment (FedericoCeratto) [reply tests/test_integration.py:71]: Unsure. It was added in ecc43915a70027cb096465e802ba3360ceadf0a1 "Make the tests work with python2"

File: tests/test_integration.py (hunk 5 above)
Comment (FedericoCeratto) [reply tests/test_integration.py:44]: It starts the container on a local port but there might be some differences.

Comment (bassosimone) [reply tests/test_integration.py:44]: Okay, thanks

File: tests/test_integration.py (hunk 6 above)
Comment (bassosimone) [reply tests/test_integration.py:48]: Maybe we can add a TODO comment saying that we can make this change in the future?

File: tests/test_integration.py (hunk 7 above)
Comment (bassosimone) [reply tests/test_integration.py:71]: I think it makes sense to document this as a py2 quirk with a comment?

File: tests/test_integration.py (hunk 8 above)
Comment (bassosimone) [reply tests/test_integration.py]: I suggest to mention this in a comment as a possible future improvement?

**Review by bassosimone — APPROVED** [review]
Now CI is working. I have left a couple more follow up comments for possible quick documentation changes that we may want to make. :whale:

File: tests/test_integration.py (hunk 7 above)
Comment (hellais) [reply tests/test_integration.py:71]: This is the relevant issue: https://github.com/docker/docker-py/issues/2255#issuecomment-475270012

Comment (bassosimone) [reply tests/test_integration.py:71]: Thanks! Can we perhaps add a comment in the code mentioning the issue?

File: tests/test_integration.py (hunk 6 above)
Comment (hellais) [reply tests/test_integration.py:48]: Added a comment.

File: tests/test_integration.py (hunk 7 above)
Comment (hellais) [reply tests/test_integration.py:71]: Added a comment for this.

**Review by hellais — APPROVED** [review]
I took a look at the diff and it looks good to me.

I committed a couple of comments directly to this branch. 

The CI is passing so for me this is good to merge.

@FedericoCeratto what was the reason for using statsd for collecting metrics vs using prometheus?

File: tests/test_integration.py (hunk 8 above)
Comment (bassosimone) [reply tests/test_integration.py]: @hellais can we also take note of this?

Comment (hellais) [reply tests/test_integration.py]: I filed an issue for it here: https://github.com/ooni/pipeline/issues/211.

**Comment by hellais** [comment]
> what was the reason for using statsd for collecting metrics vs using prometheus?

To preserve knowledge writing here what @FedericoCeratto told me privately. Basically we use Statd because it's quicker and eaiser to immediately see the metrics and gagues while developing which makes the dev cycle faster.
These are getting exposed locally using [netdata](https://github.com/netdata/netdata).

**Review by hellais — APPROVED** [review]
This looks good to me. Thanks for working on it and I can only imagine the pain doing that rebase!

**Pull request merged**